### PR TITLE
OpenCL Bindings

### PR DESCRIPTION
**Pull request by NepNet**
### Purpose of this PR

Added OpenCL bindings

### Testing status

Tested partially up to OpenCL 1.2 on Windows and Linux. Mainly getting a kernel to run and event callbacks

### CODE REVIEW

**Review by NogginBops — CHANGES_REQUESTED** [review]
For the loading of the native libraries I would like the code to be changed to be more inline with the changes and simplifications I made in https://github.com/opentk/opentk/pull/1110

**Comment by NepNet** [comment]
Updated the loading of native libraries

**Comment by NogginBops** [comment]
The updated loading looks good to me! 👍 
I haven't looked at the actual bindings, someone else will have to do that.

File: src/OpenTK.Compute/OpenCL/CL.cs [hunk 1]
```diff
@@ -0,0 +1,1728 @@
...
+		/// <param name="flags"></param>
+		/// <param name="size"></param>
+		/// <param name="hostPtr"></param>
+		/// <param name="errorCode"></param>
+		/// <returns></returns>
+		[DllImport(LibName, EntryPoint = "clCreateBuffer")]
+		public static extern IntPtr CreateBuffer(IntPtr context, MemoryFlags flags, UIntPtr size, IntPtr hostPtr,
+			out CLResultCode errorCode);
+
+		public static IntPtr CreateBuffer<T>(IntPtr context, MemoryFlags flags, T[] array, out CLResultCode errorCode)
```
Comment (jvbsl) [inline src/OpenTK.Compute/OpenCL/CL.cs:308]: I would prefer the usage of Span<T> here(for other arrays the same.

**Review by jvbsl — CHANGES_REQUESTED** [review]
Thanks for all the work.

But I would prefer stuff like platform/device being in a wrapper struct similar to ALDevice in OpenAL, would be great if you could do that

File: src/OpenTK.Compute/OpenCL/CL.cs (hunk 1 above)
Comment (NogginBops) [reply src/OpenTK.Compute/OpenCL/CL.cs:308]: At least having an overload that gives a span. Someone might have a use for an array overload

Comment (jvbsl) [reply src/OpenTK.Compute/OpenCL/CL.cs:308]: True i just recently found out that the overloading is unfortunately a little bit finicky with generics and you'd need to use CreateBuffer<float> to be able to use a Span<T> with a float[] input.
For non generics it doesn't matter though as arrays can be implicitely casted to Span ;)

**Comment by NepNet** [comment]
I give up

File: src/OpenTK.Compute/OpenCL/CL.cs [hunk 2]
```diff
@@ -313,11 +328,15 @@ public static string GetDeviceInfoAsString(IntPtr device, DeviceInfo paramName,
...
+		/// <summary>
+		/// Introduced in OpenCL 1.0
+		/// </summary>
+		public static unsafe IntPtr CreateBuffer<T>(IntPtr context, MemoryFlags flags, Span<T> span, out CLResultCode errorCode) where T : unmanaged
 		{
-			fixed (T* b = array)
+			fixed (T* b = span)
 			{
-				IntPtr buffer = CreateBuffer(context, flags, (UIntPtr)(Marshal.SizeOf<T>() * array.Length), (IntPtr)b, out errorCode);
+				IntPtr buffer = CreateBuffer(context, flags, (UIntPtr)(Marshal.SizeOf<T>() * span.Length), (IntPtr)b, out errorCode);
```
Comment (NogginBops) [inline src/OpenTK.Compute/OpenCL/CL.cs:339]: `Marshal.Sizeof<T>` can be changed to `sizeof(T)`. 
And is possibly more correct due to `Marshal.Sizeof<T>` returning the marshalled size of the struct, but there is no marshalling done.

File: tests/OpenToolkit.OpenCL.Tests/Program.cs [hunk 3]
```diff
@@ -88,9 +107,9 @@ static void Main(string[] args)
 					});
 
 					//get rid of the buffers because we no longer need them
-					CL.ReleaseMemObject(bufferA);
-					CL.ReleaseMemObject(bufferB);
-					CL.ReleaseMemObject(resultBuffer);
+					CL.ReleaseMemoryObject(bufferA);
+					CL.ReleaseMemoryObject(bufferB);
+					CL.ReleaseMemoryObject(resultBuffer);
```
Comment (varon) [inline tests/OpenToolkit.OpenCL.Tests/Program.cs:112]: We might want to name this `CL.ReleaseBuffer`.

File: src/OpenTK.Compute/OpenCL/CLEvent.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+using System;
+
+namespace OpenTK.Compute.OpenCL
+{
+	public readonly struct CLEvent : IEquatable<CLEvent>
```
Comment (varon) [inline src/OpenTK.Compute/OpenCL/CLEvent.cs:5]: Nice clean implementation here.

**Review by varon — COMMENTED** [review]
The wrapper structs are a great feature. I really like the approach of ensuring type safety through these.

Have you considered using a phantom-typed generic version of these?

Otherwise looking good, very keen to merge this as soon as we can!

**Comment by varon** [comment]
You can see an example of phantom typing here: https://github.com/opentk/opentk/pull/1016

File: tests/OpenToolkit.OpenCL.Tests/Program.cs (hunk 3 above)
Comment (NepNet) [reply tests/OpenToolkit.OpenCL.Tests/Program.cs:112]: In OpenCL Buffers, Images and Pipes are all Memory Objects, I could rename the overloads but I am not sure if it's that a good idea

File: src/OpenTK.Compute/OpenCL/CL.cs [hunk 5]
```diff
@@ -18,23 +18,27 @@ static CL()
...
 		/// <summary>
 		/// Introduced in OpenCL 1.0
+		/// See more: https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clGetPlatformIDs.html
 		/// </summary>
-		/// <param name="numberOfEntries"></param>
-		/// <param name="platforms"></param>
-		/// <param name="numberOfPlatforms"></param>
-		/// <returns></returns>
-
+		/// <param name="numberOfEntries">number of <see cref="CLPlatform"/> entries that can be added to platforms. If platforms is not NULL, numberOfEntries must be greater than zero.</param>
```
Comment (HughPH) [inline src/OpenTK.Compute/OpenCL/CL.cs:25]: Usually these start with a capital letter, e.g.
`The number of <see cref="CLPlatform"/> entries ...`

File: src/OpenTK.Compute/OpenCL/CL.cs [hunk 6]
```diff
@@ -18,23 +18,27 @@ static CL()
...
+		/// <param name="platforms">returns a list of OpenCL platforms found. The CLPlatform values returned in platforms can be used to identify a specific OpenCL platform. If platforms is NULL, this argument is ignored. The number of OpenCL platforms returned is the minimum of the value specified by numberOfEntries or the number of OpenCL platforms available.</param>
+		/// <param name="numberOfPlatforms">returns the number of OpenCL platforms available. If numberOfPlatforms is NULL, this argument is ignored.</param>
+		/// <returns>returns <see cref="CLResultCode.Success"/> if the function is executed successfully. Otherwise, it returns one of the following errors:
+		/// -InvalidValue if numberOfEntries is equal to zero and platforms is not NULL or if both numberOfPlatforms and platforms are NULL.
+		/// -OutOfHostMemory if there is a failure to allocate resources required by the OpenCL implementation on the host.</returns>
 		[DllImport(LibName, CallingConvention = CallingConvention, EntryPoint = "clGetPlatformIDs")]
 		public static extern CLResultCode GetPlatformIds([In] uint numberOfEntries, [Out] CLPlatform[] platforms,
```
Comment (HughPH) [inline src/OpenTK.Compute/OpenCL/CL.cs:32]: Is there a reason these parameters have `[In]` and `[Out]` but other methods use `in` and `out`?

Comment (NogginBops) [reply src/OpenTK.Compute/OpenCL/CL.cs:32]: There is, the in and out attributes specify how the marshaller should treat the data while the `in` and `out` keywords specify language semantics like the type being by reference (pointer).
Me and Nep figured out what was needed from the microsoft docs. The in attribute is implied but Nep wanted to be explicit about it.

Comment (HughPH) [reply src/OpenTK.Compute/OpenCL/CL.cs:32]: I support the decision to use the in attribute explicitly when the out attribute is used.

**Review by jvbsl — APPROVED** [review]

**Comment by varon** [comment]
Discussed on Discord with @NepNet  - Merging almost complete bindings! might need some work, but overall they're solid enough to get into pre!

**Comment by varon** [comment]
Discussed on Discord with @NepNet  - Merging almost complete bindings! might need some work, but overall they're solid enough to get into pre!

**Pull request merged**